python-spec/src/somacore/coordinates.py

```diff
@@ -0,0 +1,460 @@
+        if len(set(axis.name for axis in self._axes)) != len(axes):
+            raise ValueError("The names for the axes must be unique.")
+
+    def __len__(self) -> int:
+        return len(self._axes)
+
+    def __getitem__(self, index: int) -> Axis:  # type: ignore[override]
+        return self._axes[index]
+
+    def __repr__(self) -> str:
+        output = f"<{type(self).__name__}\n"
+        for axis in self._axes:
+            output += f"  {axis},\n"
+        return output + ">"
+
+    @property
+    def axes(self) -> Tuple[Axis, ...]:
+        """The axes in the coordinate space"""
+        return self._axes
+
+    @property
+    def axis_names(self) -> Tuple[str, ...]:
+        """The names of the axes in order."""
+        return tuple(axis.name for axis in self._axes)
+
+    def rank(self) -> int:
+        """The number of axes in this coordinate space."""
+        return len(self)
+
+
+class CoordinateTransform(metaclass=abc.ABCMeta):
+    """A coordinate transformation from one coordinate space to another.
+
+    Args:
+        input_axes: The names of the axes for the input coordinate space.
+        output_axes: The names of the axes for the output coordinate space.
+    """
+
+    def __init__(
+        self,
+        input_axes: Union[str, Sequence[str]],
+        output_axes: Union[str, Sequence[str]],
+    ):
+        self._input_axes = (
+            (input_axes,) if isinstance(input_axes, str) else tuple(input_axes)
+        )
+        self._output_axes = (
+            (output_axes,) if isinstance(output_axes, str) else tuple(output_axes)
+        )
+
+    @abc.abstractmethod
+    def __matmul__(self, other: Any) -> "CoordinateTransform":
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def __rmatmul__(self, other: Any) -> "CoordinateTransform":
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def inverse_transform(self) -> "CoordinateTransform":
+        """Returns the inverse coordinate transform."""
+        raise NotImplementedError()
+
+    @property
+    def input_axes(self) -> Tuple[str, ...]:
+        """The names of the axes of the input coordinate space."""
+        return self._input_axes
+
+    @property
+    def input_rank(self) -> int:
+        """The number of axes in the input coordinate space."""
+        return len(self._input_axes)
+
+    @property
+    def output_axes(self) -> Tuple[str, ...]:
+        """The names of the axes of the output coordinate space"""
+        return self._output_axes
+
+    @property
+    def output_rank(self) -> int:
+        """The number of axes in the output coordinate space."""
+        return len(self._output_axes)
+
+
+class AffineTransform(CoordinateTransform):
+    """An affine coordinate trasformation from one coordinate space to another.
+
+    An affine transform is a combination of a linear transformation and a translation.
+    For a vector x it can be written as
+
+        y = Ax + b
```

Comment:
This is a little confusing. In this line you're showing `A` as non-augmented but below you are describing it in augmented form. Maybe just a sentence here saying the "+ b" part can be folded into the matrix using the standard augmented-matrix trick ...